### PR TITLE
Add trigger response types, accept workflows as a generic

### PR DESCRIPTION
**Pull request by shapirone**
##  Summary

This PR introduces two features and a ton of tests
1. Trigger methods now receive workflows as an optional generic. This provides typing for the `workflow` field and it powers the `inputs` field. It's used by both CLI files and at runtime.
2. Triggers now have response typing. This is only used at runtime.

```ts
// manifest.ts
export const MixedWorkflow = DefineWorkflow({
  callback_id: "mixed",
  title: "Example",
  input_parameters: {
    properties: {
      optional: {
        type: "string",
      },
      required: {
        type: "string",
      },
    },
    required: ["required"],
  },
});

// reverse.ts
const resp = await client.workflows.triggers.create<typeof MixedWorkflow.definition>({
  inputs: {} // <- This should now be typed
  workflow: // <- This should now be typed 
})

if (resp.ok) {
  resp.trigger; // <- This should now be typed
}
```

#### CLI
```ts
const trigger: Trigger<typeof ReverseWorkflow.definition> = {
  name: "Test Trigger",
  type: "shortcut",
  workflow: "#/workflows/reverse_wf",
};
```
#### Runtime
```ts
await client.workflows.triggers.create<typeof ReverseWorkflow.definition> = {
  name: "Test Trigger",
  type: "shortcut",
  workflow: "#/workflows/reverse_wf",
}
```

#### Testing

Since these are just type changes testing should mostly be done against the experience to write triggers, but it's always good to make sure things are working e2e
* Test with and without generics
* Test workflows that require inputs, workflows that don't have any inputs, workflows that have optional inputs, etc.
* Test to make sure that `workflow` param is always right
* Verify the response is what you'd expect for the different trigger types
* Try to find any case where we're not providing flexibility where we should

### Requirements (place an `x` in each `[ ]`)

* [x] I've read and understood the [Contributing Guidelines](https://github.com/slackapi/deno-slack-api/blob/main/.github/CONTRIBUTING.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).

### CODE REVIEW

File: src/typed-method-types/workflows/triggers/event.ts [hunk 1]
```diff
@@ -57,6 +60,8 @@ type ChannelEvents =
   & {
     /** @description The channel id's that this event listens on */
     channel_ids: PopulatedArray<string>;
+    // deno-lint-ignore no-explicit-any
+    [otherOptions: string]: any;
```
Comment (shapirone) [inline src/typed-method-types/workflows/triggers/event.ts:64]: These are intentionally strewn across trigger inputs in the PR. The intent is to make sure that if triggers start to accept new data that we don't expect, they are still usable without typeahead.

File: src/typed-method-types/workflows/triggers/inputs.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    & { [k in Params["required"][number]]: WorkflowInput }
+  : Record<string, WorkflowInput>;
+
+/**
+ * Determines if the input object itself is required to pass to the trigger or not.
+ * Returns an object where the input object is either required or optional.
+ */
+type WorkflowInputsType<Params extends InputParameterSchema> =
+  // This intentionally avoids Distributive Conditional Types, so be careful removing any of the following square brackets
+  // See https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#distributive-conditional-types for more details
```
Comment (shapirone) [inline src/typed-method-types/workflows/triggers/inputs.ts:47]: This bit was the doozy that held up the PR for so long

File: src/typed-method-types/workflows/triggers/mod.ts [hunk 3]
```diff
@@ -11,64 +28,122 @@ export const TriggerTypes = {
...
+  trigger?: never;
+};
+
+type TriggerIdType = {
+  /** @description The id of a specified trigger */
+  trigger_id: string;
+};
+
+export type ValidTriggerTypes<
+  WorkflowDefinition extends WorkflowSchema = DEFAULT_WORKFLOW_TYPE,
```
Comment (shapirone) [inline src/typed-method-types/workflows/triggers/mod.ts:112]: This default allows workflow generics to not be passed while still making sure that inputs aren't treated as not existing.

File: src/typed-method-types/workflows/triggers/base_response.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    /** @description The trigger's internal uid */
+    id: string;
+    /** @description The metadata for the workflow */
+    workflow: {
+      /** @description the workflow function's internal uid */
+      id: string;
+      /** @description the workflow's internal uid */
+      workflow_id: string;
+      /** @description the workflow's callback_id */
+      callback_id: string;
```
Comment (filmaj) [inline src/typed-method-types/workflows/triggers/base_response.ts:17]: Can some of these properties, like callback_id and title, be narrowed further to the exact value as defined in the WorkflowDefinition? I.e. set this to something like 
```suggestion
      callback_id: WorkflowDefinition['callback_id'];
```

File: src/typed-method-types/workflows/triggers/inputs.ts (hunk 2 above)
Comment (filmaj) [reply src/typed-method-types/workflows/triggers/inputs.ts:47]: motherofgod

**Review by filmaj — APPROVED** [review]
walkenthanksyou

Thanks for tackling this! Works great. And thank you for TypeScript expertise and sacrifice 🙏 

Left a question, minor, mostly for my own education I think.

File: src/typed-method-types/workflows/triggers/base_response.ts (hunk 4 above)
Comment (shapirone) [reply src/typed-method-types/workflows/triggers/base_response.ts:17]: Yes they can. IMO the value of it is questionable. For instance, in practice this would mean that when the developer gets a response, the `triggerResponse.trigger.workflow.callback_id` will be typed to the specific string that they already provided above, or that `triggerResponse.trigger.workflow.inputs` provides typeahead for its available properties. I don't see much of a point to that personally, but if you have a use case in mind I'm open to changing my mind there.

TLDR — We can do this if we want, but value is limited and I'm going to say it's outside of the scope of this PR

Comment (filmaj) [reply src/typed-method-types/workflows/triggers/base_response.ts:17]: Sounds good to me!
